module/zfs/dmu_recv.c

```diff
@@ -1269,13 +1269,21 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
 	if (payloadlen != 0)
 		payload = vmem_alloc(payloadlen, KM_SLEEP);
 
-	err = receive_read_payload_and_next_header(drc, payloadlen,
-	    payload);
-	if (err != 0) {
-		vmem_free(payload, payloadlen);
-		return (err);
-	}
 	if (payloadlen != 0) {
```

Comment:
We now have two `if (payloadlen != 0)` checks next to each other, let's pull everything together under one.  It looks like we can pull the variable declaration for `payload` in to this block as well.